tools/src/wyvern/tools/parsing/coreparser/arch/ASTComponentTypeDecl.java

```diff
@@ -0,0 +1,175 @@
+import wyvern.target.corewyvernIL.decl.Declaration;
+import wyvern.target.corewyvernIL.decl.DefDeclaration;
+import wyvern.target.corewyvernIL.decltype.DeclType;
+import wyvern.target.corewyvernIL.decltype.ValDeclType;
+import wyvern.target.corewyvernIL.expression.Expression;
+import wyvern.target.corewyvernIL.expression.New;
+import wyvern.target.corewyvernIL.modules.Module;
+import wyvern.target.corewyvernIL.support.InterpreterState;
+import wyvern.target.corewyvernIL.type.NominalType;
+import wyvern.target.corewyvernIL.type.StructuralType;
+import wyvern.tools.errors.ErrorMessage;
+import wyvern.tools.errors.ToolError;
+
+public class ASTComponentTypeDecl extends SimpleNode {
+    private boolean isExternal = false;
+    private String typeName;
+    private HashMap<String, String> reqs = new HashMap<>();
+    private HashMap<String, String> provs = new HashMap<>();
+
+    public ASTComponentTypeDecl(int id) {
+        super(id);
+    }
+
+    public ASTComponentTypeDecl(ArchParser p, int id) {
+        super(p, id);
+    }
+
+    public HashMap<String, String> getReqs() {
+        return reqs;
+    }
+
+    public HashMap<String, String> getProvs() {
+        return provs;
+    }
+
+    public boolean isExternal() {
+        return isExternal;
+    }
+
+    public void setExternal(boolean isExternal) {
+        this.isExternal = isExternal;
+    }
+
+    public String getTypeName() {
+        return typeName;
+    }
+
+    public void setTypeName(String t) {
+        typeName = t;
+    }
+
+    public String toString() {
+        return super.toString() + " " + typeName + " : " + isExternal;
+    }
+
+    public void collectPorts() {
+        for (Node child : this.getChildren()) {
+            if (child instanceof ASTPortDecl) {
+                String req = ((ASTPortDecl) child).getRequires();
+                String prov = ((ASTPortDecl) child).getProvides();
+                String name = ((ASTPortDecl) child).getPort();
+                if (req != "") {
+                    reqs.put(name, req);
+                } else if (prov != "") {
+                    provs.put(name, prov);
+                }
+            }
+        }
+    }
+
+    public boolean checkModule(InterpreterState state) {
+        if (isExternal) {
+            return true;
+        }
+        Module mod = null;
+        int checkCount = 0;
+        try {
+            mod = state.getResolver().resolveModule(typeName);
+
+            /* Check that it's a module def */
+            Expression expr = mod.getExpression();
+            if (!expr.getType().toString().contains("apply(")) {
+                System.out.println(expr.getType().toString());
+                // throw error saying it's not a module def
+                ToolError.reportError(ErrorMessage.MODULE_DEF_NOT_FOUND, this.getLocation(),
+                        typeName);
+            }
+
+            /* Check for only for dependencies passed as args to module def */
+            if (expr instanceof New) {
```

Comment:
expr has to be a module def. Please remove the if statement and replace it with an assert. Do the check for dependencies after the assert.